IBPSA/Fluid/Actuators/Valves/TwoWayPressureIndependent.mo

```diff
@@ -12,69 +12,115 @@ model TwoWayPressureIndependent "Model of a pressure-independent two way valve"
-                                  m_flow_turbulent=m_flow_turbulent),
-                            x_small=dp_nominal_pos*deltax);
-      else
-        dp=IBPSA.Utilities.Math.Functions.regStep(
-                            x=m_flow-m_flow_set,
-                            y1= dp_min + (m_flow-m_flow_set)/m_flow_nominal*dp_nominal/l2,
-                            y2= IBPSA.Fluid.BaseClasses.FlowModels.basicFlowFunction_m_flow(
-                                  m_flow=m_flow,
-                                  k=k,
-                                  m_flow_turbulent=m_flow_turbulent),
-                            x_small=m_flow_nominal_pos*deltax*l2);
-      end if;
-    end if; // homotopyInitialization
+  // C2 smooth transition between y1 and y2
+  y_smooth = noEvent(smooth(2,
+        if x <= x1
+        then y1
+        elseif x >=x2
+        then y2
+        else IBPSA.Utilities.Math.Functions.quinticHermite(
+                 x=x,
+                 x1=x1,
+                 x2=x2,
+                 y1=y1,
+                 y2=y2,
+                 y1d= if from_dp
+                      then IBPSA.Fluid.BaseClasses.FlowModels.basicFlowFunction_dp_der(
+                                     dp=dp_min + x1,
+                                     k=k,
+                                     m_flow_turbulent=m_flow_turbulent,
+                                     dp_der=1)
+                      else IBPSA.Fluid.BaseClasses.FlowModels.basicFlowFunction_m_flow_der(
+                                     m_flow=m_flow_set + x1,
+                                     k=k,
+                                     m_flow_turbulent=m_flow_turbulent,
+                                     m_flow_der=1),
+                 y2d=y2d,
+                 y1dd=if from_dp
+                      then IBPSA.Fluid.BaseClasses.FlowModels.basicFlowFunction_dp_der2(
+                                     dp=dp_min + x1,
+                                     k=k,
+                                     m_flow_turbulent=m_flow_turbulent,
+                                     dp_der=1,
+                                     dp_der2=0)
+                      else IBPSA.Fluid.BaseClasses.FlowModels.basicFlowFunction_m_flow_der2(
+                                     m_flow=m_flow_set + x1,
+                                     k=k,
+                                     m_flow_turbulent=m_flow_turbulent,
+                                     m_flow_der=1,
+                                     m_flow_der2=0),
+                 y2dd=y2dd)));
+
+  if homotopyInitialization then
+    if from_dp then
+      m_flow=homotopy(actual=y_smooth,
+                      simplified=m_flow_nominal_pos*dp/dp_nominal_pos);
+    else
+        dp=homotopy(
+           actual=y_smooth,
+           simplified=dp_nominal_pos*m_flow/m_flow_nominal_pos);
+    end if;
+  else
+    if from_dp then
+      m_flow=y_smooth;
+    else
+      dp=y_smooth;
+    end if;
+  end if;
```

Comment:
`y_smooth = if from_dp then m_flow else dp`, hence the units of `y_smooth` change depending on the parameter `from_dp`. I don't know if this is valid Modelica and hence suggest to change it so that you have a different variable for each case. Then, you can specify its units when you declare the variable.